packages/components/src/spinner/index.tsx

```diff
@@ -7,19 +7,12 @@ import classNames from 'classnames';
  * Internal dependencies
  */
 import { StyledSpinner, SpinnerTrack, SpinnerIndicator } from './styles';
+import type { SpinnerProps } from './types';
 
-/**
- * @typedef OwnProps
- *
- * @property {string} [className] Class name
- */
-/** @typedef {import('react').ComponentPropsWithoutRef<'svg'> & OwnProps} Props */
-
-/**
- * @param {Props} props
- * @return {JSX.Element} Element
- */
-export default function Spinner( { className, ...props } ) {
+export default function Spinner( {
+	className,
+	...props
+}: SpinnerProps ): JSX.Element {
```

Comment:
If we use the `WordPressComponentProps` utility type instead, we can avoid specifying the `className` prop (and therefore, adding the `types.ts` file altogether):

```diff
diff --git a/packages/components/src/spinner/index.tsx b/packages/components/src/spinner/index.tsx
index f1493a6c1c..ebe47628a2 100644
--- a/packages/components/src/spinner/index.tsx
+++ b/packages/components/src/spinner/index.tsx
@@ -7,12 +7,12 @@ import classNames from 'classnames';
  * Internal dependencies
  */
 import { StyledSpinner, SpinnerTrack, SpinnerIndicator } from './styles';
-import type { SpinnerProps } from './types';
+import type { WordPressComponentProps } from '../ui/context';
 
 export default function Spinner( {
 	className,
 	...props
-}: SpinnerProps ): JSX.Element {
+}: WordPressComponentProps< {}, 'svg', false > ) {
 	return (
 		<StyledSpinner
 			className={ classNames( 'components-spinner', className ) }
diff --git a/packages/components/src/spinner/types.ts b/packages/components/src/spinner/types.ts
deleted file mode 100644
index 596dafdb7a..0000000000
--- a/packages/components/src/spinner/types.ts
+++ /dev/null
@@ -1,3 +0,0 @@
-export interface SpinnerProps extends React.ComponentPropsWithoutRef< 'svg' > {
-	className: string;
-}
```

The `WordPressComponentProps< {}, 'svg', false > )` expression means:

- use the `WordPressComponentProps` type
- don't add any extra props (hence the `{}`)
- add all props and attributes of the `svg` element (including `className`)
- mark the component as non-polymorphic by not accepting the `as` prop (hence the `false`), since it should always be rendered as an SVG


We also don't need to explicitly add `JSX.Element` as a return value.

Reply:
Thanks for taking the time to explain this @ciampo this information is really valuable. I'll take this into account on future PRs.

> We also don't need to explicitly add JSX.Element as a return value.

Good to know, thank you :)